### PR TITLE
bsd: clear_user_and_ssh_keys script should not fail if metadata does not exist

### PR DESCRIPTION
**Pull request by nbyavuz**
clear_user_and_ssh_keys script does not fail when the GCP_PROJECT is 'pg-ci-images-dev' but fails when it is 'pg-ci-images'. Most possible reason is 'pg-ci-images' project does not have at least one of the project_level_ssh_keys and instance_level_ssh_keys. So as a solution, script should not fail if it could not find these keys.

### CODE REVIEW

**Comment by anarazel** [comment]
That seems likely to hide bugs. Could we just ignore 404s or such?

**Comment by nbyavuz** [comment]
> That seems likely to hide bugs. Could we just ignore 404s or such?

I could not find a way to directly check if there is 404, so I used grep for that.

File: scripts/bsd/clear_users_and_ssh_keys.sh [hunk 1]
```diff
@@ -19,9 +19,24 @@ fi
...
-project_level_ssh_keys=$(curl -fsH "Metadata-Flavor: Google"  http://metadata.google.internal/computeMetadata/v1/project/attributes/ssh-keys)
-instance_level_ssh_keys=$(curl -fsH "Metadata-Flavor: Google"  http://metadata.google.internal/computeMetadata/v1/instance/attributes/ssh-keys)
-ssh_keys=$(printf '%s\n%s' "$project_level_ssh_keys" "$instance_level_ssh_keys")
+ssh_keys=""
+
+# do not fail if curl fails with 404
+curl_helper ()
+{
+    exit_code=0
+    curl_output=$(curl -fsSH "Metadata-Flavor: Google" $1 2>&1) || exit_code=$?
```
Comment (anarazel) [inline scripts/bsd/clear_users_and_ssh_keys.sh:28]: Why not just do this portion unconditionally? I.e.
```bash
    curl_output=$(curl -fsSH "Metadata-Flavor: Google" $1 2>&1)
    exit_code=$?
```

File: scripts/bsd/clear_users_and_ssh_keys.sh [hunk 2]
```diff
@@ -19,9 +19,24 @@ fi
...
+    if [ "${exit_code}" != 0 ] && (echo ${curl_output} | grep -v "404")
+    then
+        echo "Curl failed with exit code ${exit_code}"
+        exit ${exit_code}
+    else
+        ssh_keys=$(printf '%s\n%s' "$ssh_keys" "$curl_output")
```
Comment (anarazel) [inline scripts/bsd/clear_users_and_ssh_keys.sh:34]: Hm, is this actually going to print something sensible in the 404 case, given that you're doing `2>&1`? I'd probably just add a dedicated branch for the 404 situation.

File: scripts/bsd/clear_users_and_ssh_keys.sh (hunk 1 above)
Comment (anarazel) [reply scripts/bsd/clear_users_and_ssh_keys.sh:28]: Hm, actually, there might be a better option:
```bash
curl --fail --silent -o /tmp/test -w "%{http_code}\n" 'https://anarazel.de/nofile'
404
```
```bash
curl --fail --silent -o /tmp/test -w "%{http_code}\n" 'https://anarazel.de/'
200
```

Comment (nbyavuz) [reply scripts/bsd/clear_users_and_ssh_keys.sh:28]: > Hm, actually, there might be a better option:
> 
> ```shell
> curl --fail --silent -o /tmp/test -w "%{http_code}\n" 'https://anarazel.de/nofile'
> 404
> ```
> 
> ```shell
> curl --fail --silent -o /tmp/test -w "%{http_code}\n" 'https://anarazel.de/'
> 200
> ```

That looks cleaner but then curl's output needs to be gotten from `/tmp/test` file and this file needs to be removed as well. Would you prefer that approach?

File: scripts/bsd/clear_users_and_ssh_keys.sh (hunk 2 above)
Comment (nbyavuz) [reply scripts/bsd/clear_users_and_ssh_keys.sh:34]: > Hm, is this actually going to print something sensible in the 404 case, given that you're doing `2>&1`? I'd probably just add a dedicated branch for the 404 situation.

I am not sure. -f flag means there will be no body output if curl fails. But I agree with that it is better to add another branch for this case.

File: scripts/bsd/clear_users_and_ssh_keys.sh (hunk 1 above)
Comment (anarazel) [reply scripts/bsd/clear_users_and_ssh_keys.sh:28]: Yes - I'd not use /tmp/test though. Use `mktemp` or such.

**Review by anarazel — APPROVED** [review]

**Pull request merged**